rmc/src/handler.rs

```diff
@@ -0,0 +1,118 @@
+//! Reliable MultiCast - a primitive for Reliable Broadcast protocol.
+pub use aleph_bft_crypto::{
+    Indexed, MultiKeychain, Multisigned, NodeCount, PartialMultisignature, PartiallyMultisigned,
+    Signable, Signature, Signed, UncheckedSigned,
+};
+use core::fmt::Debug;
+use std::{
+    collections::HashMap,
+    fmt::{Display, Formatter},
+    hash::Hash,
+};
+
+pub enum Error {
+    BadSignature,
+    BadMultisignature,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::BadSignature => write!(f, "received a hash with a bad signature."),
+            Error::BadMultisignature => write!(f, "received a hash with a bad multisignature."),
+        }
+    }
+}
+
+pub struct Handler<H: Signable + Hash, MK: MultiKeychain> {
+    keychain: MK,
+    hash_states: HashMap<H, PartiallyMultisigned<H, MK>>,
+}
+
+impl<H: Signable + Hash + Eq + Clone + Debug, MK: MultiKeychain> Handler<H, MK> {
+    pub fn new(keychain: MK) -> Self {
+        Handler {
+            hash_states: HashMap::new(),
+            keychain,
+        }
+    }
+
+    /// Signs hash and updates the internal state with it. Returns the unchecked signed
+    /// version of the hash for broadcast. Should be called at most once for a particular hash.
+    pub fn on_start_rmc(&mut self, hash: H) -> UncheckedSigned<Indexed<H>, MK::Signature> {
+        let unchecked = Signed::sign_with_index(hash.clone(), &self.keychain).into_unchecked();
+        let _ = self.on_signed_hash(&unchecked);
+        unchecked
+    }
+
+    /// Update the internal state with the signed hash. If the hash is incorrectly signed then
+    /// [`Error::BadSignature`] is returned. If Adding this signature completes a multisignature
+    /// then `Ok(multisigned)` is returned. Otherwise `Ok(None)` is returned.
+    pub fn on_signed_hash(
+        &mut self,
+        unchecked: &UncheckedSigned<Indexed<H>, MK::Signature>,
```

Comment:
`let hash = ...` makes a borrow, so cloning below is still needed. When keeping the reference in the argument, at least we save one clone, cuz we can call `on_signed_hash(&x)` instead `on_signed_hash(x.clone())`

Reply:
Also it is consistent with other methods & outer scope functions which take `unchecked` by reference

Reply:
Hmm, yeah, I kinda like the idea of separating `handle_signed_hash` out so that we don't check the signature produced by us in `on_start_rmc`. However your implementation doesn't seem to work for me (even after some adjustments). 

`handle_signed_hash` takes `Signed` instead of `UncheckedSigned` which is reasonable as we don't want to check the signature, but `UncheckedSigned` is **needed** here as extracting the `hash` is only possible through `UncheckedSigned` (by calling `as_signable_strip_index`). `Signed` can only be converted to `UncheckedSigned` by consuming the original `Signed` (`into_unchecked`). Thus some cloning is inevitable here (if we want to pass `signed` to `add_signature` later) :smile:. 

Although, I think that the implementation of `Signed` and `Indexed` is unnecessarily restricted. Why do we have to convert to `UncheckedSigned` each time we want to access the `hash` directly? I think that hash residing in `Indexed` should be accessible by reference (e.g. by introducing `as_signable(&self) -> &T` in `Indexed`) and it would allow accessing the `hash` without the need of cloning `Signed` to convert to `UncheckedSigned` and then to access the hash.

Here is how I see `handle_signed_hash` now:
```rust
fn handle_signed_hash(&mut self, signed: Signed<Indexed<H>, MK>) -> Option<Multisigned<H, MK>> {
        let unchecked = signed.clone().into_unchecked(); // needa clone signed to get the hash
        let hash = unchecked.as_signable_strip_index();
        if self.already_completed(hash) {
            return None;
        }
        let new_state = match self.hash_states.remove(hash) {
            None => signed.into_partially_multisigned(&self.keychain),
            Some(partial) => partial.add_signature(signed, &self.keychain),
        };
        match new_state {
            PartiallyMultisigned::Complete { multisigned } => {
                self.hash_states.insert(
                    hash.clone(),
                    PartiallyMultisigned::Complete {
                        multisigned: multisigned.clone(),
                    },
                );
                Some(multisigned)
            }
            incomplete => {
                self.hash_states.insert(hash.clone(), incomplete);
                None
            }
        }
    }
```
and here is how it could look like if we could access &hash from `Indexed`:
```rust
fn handle_signed_hash(&mut self, signed: Signed<Indexed<H>, MK>) -> Option<Multisigned<H, MK>> {
        let hash = signed.as_signable().as_signable().clone(); // access hash without cloning the signed
        if self.already_completed(&hash) {
            return None;
        }
        let new_state = match self.hash_states.remove(&hash) {
            None => signed.into_partially_multisigned(&self.keychain),
            Some(partial) => partial.add_signature(signed, &self.keychain),
        };
        match new_state {
            PartiallyMultisigned::Complete { multisigned } => {
                self.hash_states.insert(
                    hash.clone(),
                    PartiallyMultisigned::Complete {
                        multisigned: multisigned.clone(),
                    },
                );
                Some(multisigned)
            }
            incomplete => {
                self.hash_states.insert(hash.clone(), incomplete);
                None
            }
        }
    }
```